app/common/PetscLinearSolver.h

```diff
@@ -0,0 +1,59 @@
+#ifndef PETSCSOLVER_20200910_H
+#define PETSCSOLVER_20200910_H
+
+#include "PetscUtil.h"
+#include "common/PetscBlockMatrix.h"
+#include "common/PetscBlockVector.h"
+
+#include <petscksp.h>
+#include <petscsys.h>
+#include <petscsystypes.h>
+
+#include <cstddef>
+#include <memory>
+#include <utility>
+
+namespace tndm {
+
+class PetscLinearSolver {
+public:
+    template <typename DGOp> PetscLinearSolver(DGOp& dgop) {
+        auto const& topo = dgop.topo();
+        A_ = std::make_unique<PetscBlockMatrix>(dgop.block_size(), topo.numLocalElements(),
+                                                topo.numLocalNeighbours(),
+                                                topo.numGhostNeighbours(), topo.comm());
+        b_ = std::make_unique<PetscBlockVector>(dgop.block_size(), topo.numLocalElements(),
+                                                topo.comm());
+        x_ = std::make_unique<PetscBlockVector>(*b_);
+        dgop.assemble(*A_);
+        dgop.rhs(*b_);
+
+        CHKERRTHROW(KSPCreate(topo.comm(), &ksp_));
```

Comment:
@uphoffc It would be very useful to allow one to attach prefix to the KSP via KSPSetOptionsPrefix() here. Thus enabling command line args to define the configuration of a specific KSP object. -poisson_ksp_type cg -elasticity_ksp_type gmres